Assets/Samples/IMEInput/IMEInputHandler.cs

```diff
@@ -0,0 +1,136 @@
+            {
+                m_CompositionField.text = string.Empty;
+                m_ComposingViaIME = false;
+            }
+
+            m_CurrentInput += character;
+            m_TextField.text = m_CurrentInput;
+            m_CombinedField.text = m_CurrentInput;
+        }
+
+        // Handles text passed via Keyboard.onIMECompositionChange event
+        private void OnIMECompositionChange(IMECompositionString text)
+        {
+            AssertInvariance();
+
+            // IME composition strings without length can also mean
+            // the composition has been submitted
+            if (text.Count == 0)
+            {
+                m_ComposingViaIME = false;
+                m_CompositionField.text = string.Empty;
+                return;
+            }
+
+            var compositionText = text.ToString();
+            m_ComposingViaIME = true;
+            m_CompositionField.text = compositionText;
+
+            // The combined text contains both the current input and the current status of the composition
+            m_CombinedField.text = string.Format("{0}{1}", m_CurrentInput, compositionText);
+        }
+
+        // Adds keyboard and input field listeners
+        private void OnEnable()
+        {
+            if (m_EventListenersAdded)
+                return;
+
+            var keyboard = InputSystem.GetDevice<Keyboard>();
+            if (keyboard is null)
+                return;
+
+            keyboard.onTextInput += OnTextInput;
+            keyboard.onIMECompositionChange += OnIMECompositionChange;
+            m_InputField.onValueChanged.AddListener(OnValueChanged);
+
+            m_EventListenersAdded = true;
+        }
+
+        // Removes keyboard and input field listeners
+        private void OnDisable()
+        {
+            if (!m_EventListenersAdded)
+                return;
+
+            var keyboard = InputSystem.GetDevice<Keyboard>();
+            if (keyboard is null)
+                return;
+
+            keyboard.onTextInput -= OnTextInput;
+            keyboard.onIMECompositionChange -= OnIMECompositionChange;
+            m_InputField.onValueChanged.RemoveListener(OnValueChanged);
+
+            Clear();
+
+            m_EventListenersAdded = false;
+        }
+
+        // Called when the input field's text is changed
+        private void OnValueChanged(string value)
+        {
+            AssertInvariance();
+
+            if (!string.IsNullOrEmpty(value))
+                return;
+
+            Clear();
+        }
+
+        // Clears the text from all of the fields
+        private void Clear()
+        {
+            m_CompositionField.text = string.Empty;
+            m_TextField.text = string.Empty;
+            m_CombinedField.text = string.Empty;
+
+            m_CurrentInput = string.Empty;
+            m_ComposingViaIME = false;
+        }
+
+        // Ensures all fields are correctly referenced
+        private void AssertInvariance()
```

Comment:
Strange naming but I like the check

Reply:
I agree actually, maybe `EnsureReferencesNotNull` or something like that would be clearer. Will get this switched out today.